services/src/main/java/org/keycloak/authentication/authenticators/browser/RecoveryAuthnCodesFormAuthenticator.java

```diff
@@ -0,0 +1,153 @@
+                    RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW));
+        } else {
+            targetRealm = authnFlowContext.getRealm();
+            authenticatedUser = authnFlowContext.getUser();
+            if (!isDisabledByBruteForce(authnFlowContext, authenticatedUser)) {
+                isValid = this.userCredentialManager.isValid(targetRealm, authenticatedUser,
+                        UserCredentialModel.buildFromBackupAuthnCode(recoveryAuthnCodeUserInput.replace("-", "")));
+                if (!isValid) {
+                    responseChallenge = createLoginForm(authnFlowContext, true,
+                            RecoveryAuthnCodesUtils.RECOVERY_AUTHN_CODES_INPUT_DEFAULT_ERROR_MESSAGE,
+                            RecoveryAuthnCodesUtils.FIELD_RECOVERY_CODE_IN_BROWSER_FLOW);
+                    authnFlowContext.failureChallenge(AuthenticationFlowError.INVALID_CREDENTIALS, responseChallenge);
+                } else {
+                    result = true;
+                    optUserCredentialFound = this.userCredentialManager.getStoredCredentialsByTypeStream(targetRealm,
+                            authenticatedUser, RecoveryAuthnCodesCredentialModel.TYPE).findFirst();
+                    if (optUserCredentialFound.isPresent()) {
+                        recoveryCodeCredentialModel = RecoveryAuthnCodesCredentialModel
+                                .createFromCredentialModel(optUserCredentialFound.get());
+                        if (recoveryCodeCredentialModel.allCodesUsed()) {
+                            this.userCredentialManager.removeStoredCredential(targetRealm, authenticatedUser,
+                                    recoveryCodeCredentialModel.getId());
+                        }
+                    }
+                    if (recoveryCodeCredentialModel == null || recoveryCodeCredentialModel.allCodesUsed()) {
+                        authenticatedUser.addRequiredAction(UserModel.RequiredAction.CONFIGURE_RECOVERY_AUTHN_CODES);
+                    }
+                }
+            }
+        }
+        return result;
+    }
+
+    protected boolean isDisabledByBruteForce(AuthenticationFlowContext authnFlowContext, UserModel authenticatedUser) {
+        boolean bolResult = false;
+        String bruteForceError;
+        Response challengeResponse;
+        bruteForceError = AuthenticatorUtils.getDisabledByBruteForceEventError(authnFlowContext.getProtector(),
+                authnFlowContext.getSession(), authnFlowContext.getRealm(), authenticatedUser);
+        if (bruteForceError != null) {
+            authnFlowContext.getEvent().user(authenticatedUser);
+            authnFlowContext.getEvent().error(bruteForceError);
+            challengeResponse = createLoginForm(authnFlowContext, false, Messages.INVALID_USER, FIELD_USERNAME);
+            authnFlowContext.forceChallenge(challengeResponse);
+            bolResult = true;
+        }
+        return bolResult;
+    }
```

Comment:
@mposolda does an authenticator have to deal with brute-force itself? I would have thought that would be built-into the flow somehow?

Reply:
I was thinking the same, but it is not :/ The individual authenticators currently checks by itself (For example `AbstractUsernameFormAuthenticator.isDisabledByBruteForce` ).

I don't know what was the motivation to do it this way? I guess it can be related to the fact that you still want to be able to authenticate user for example by `CookieAuthenticator` (or some others) even if he is temporarily disabled.

So for example flow like:
- Valid user authenticates and creates session1
- Attacker tries brute forcing in different laptop and this cause user to be temporarily disabled
- Valid user is currently still able to SSO re-authenticate in session1 even if he is disabled by brute force.
 
If we check brute force at the flow level for every authenticator, it would mean that SSO re-authentication with cookie may not work and hence attacker has "power" to break the valid SSO sessions of existing users. Which is also not ideal...

Reply:
Is there consensus on whether we should leave it as is, or remove the brute-force code?

Reply:
We should leave it as is since there's no "global" check for it. @mposolda I think it may be worth re-visiting in the future, so it's not individual authenticators that control brute-force, but I do agree that the "cookie" is a special case though

Reply:
@stianst "Neither admin or account console shows how many codes are left" the account console shows how many codes are left. Where would we show it on the admin console? Is that something we can revisit later?